### PR TITLE
Add configurable logging to the console and support stdout sinks

### PR DESCRIPTION
**Pull request by adfoster-r7**
As an alternative to the rather specific implementation of [Add stdout logging functionality to json rpc](https://github.com/rapid7/metasploit-framework/pull/14377), this PR adds a configuration option to the console and the rpc service for specifying the required metasploit logsink

### Console verification

Verify that default logging works, and the default Flatfile sink works as expected:
```
bundle exec ./msfconsole
cat ~/.msf4/logs/framework.log
```

Ensure that the logger can be configured with arbitrary loggers:
```
bundle exec ./msfconsole --logger Stdout -x 'exit' -q
bundle exec ./msfconsole --logger StdoutWithoutTimestamps -x 'exit' -q
bundle exec ./msfconsole --logger Stderr -x 'exit' -q
bundle exec ./msfconsole --logger Flatfile -x 'exit' -q
bundle exec ./msfconsole --logger TimestampColorlessFlatfile -x 'exit' -q
bundle exec ./msfconsole --logger LoggerThatDoesntExist -x 'exit' -q
```

For example, the stdout/stderr loggers should resemble:
```
$ bundle exec ./msfconsole --logger Stdout -q
[11/25/2020 17:54:12] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[11/25/2020 17:54:12] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[11/25/2020 17:54:12] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[11/25/2020 17:54:12] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[11/25/2020 17:54:13] [e(0)] core: Unable to load module /Users/adfoster/Documents/code/metasploit-framework/modules/auxiliary/gather/office365userenum.py - LoadError  Try running file manually to check for errors or dependency issues.
[*] Using configured payload php/meterpreter/reverse_tcp
[*] Starting persistent handler(s)...
msf6 exploit(multi/http/drupal_drupageddon) > 
```

The Flatfile loggers will go to `~/.msf4/logs/framework.log` as normal.

Additionally verifying invalid options work:

```
$ bundle exec ./msfconsole --logger InvalidLogger -x 'exit' -q
bundler: failed to load command: ./msfconsole (./msfconsole)
Rex::ArgumentError: An invalid argument was specified. Could not find logger InvalidLogger, expected one of StdoutWithoutTimestamps, FlatfileWithoutColors, Stderr, Stdout, Flatfile, Stream
...
```

### JSON RPC support

This functionality is additionally supported by the rpc layer:
```
MSF_WS_DATA_SERVICE_LOGGER=Stdout bundle exec thin --rackup msf-json-rpc.ru --address localhost --port 8081 --environment production --tag msf-json-rpc start 
```

Curling with an additional request:
```
curl --request POST --url http://localhost:8081/api/v1/json-rpc --header 'content-type: application/json' --data '{ "jsonrpc": "2.0", "method": "core.version", "id": 1, "params": [] }'
```

Verifying the server's output:
```
$ MSF_WS_DATA_SERVICE_LOGGER=Stdout bundle exec thin --rackup msf-json-rpc.ru --address localhost --port 8081 --environment production --tag msf-json-rpc start
[11/25/2020 17:34:53] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[11/25/2020 17:34:53] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[11/25/2020 17:34:53] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[11/25/2020 17:34:53] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[11/25/2020 17:34:54] [e(0)] core: Unable to load module /Users/adfoster/Documents/code/metasploit-framework/modules/auxiliary/gather/office365userenum.py - LoadError  Try running file manually to check for errors or dependency issues.
Thin web server (v1.7.2 codename Bachmanity)
Maximum connections set to 1024
Listening on localhost:8081, CTRL+C to stop
[11/25/2020 17:35:17] [d(0)] core: Already established connection to postgresql, so reusing active connection.
[11/25/2020 17:35:17] [e(0)] core: DB.connect threw an exception - ActiveRecord::AdapterNotSpecified database configuration does not specify adapter
[11/25/2020 17:35:17] [e(0)] core: Failed to connect to the database: database configuration does not specify adapter
```

### CODE REVIEW

File: lib/rex/logging/sinks/stream.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    when LOG_ERROR
+      code = 'e'
+    when LOG_INFO
+      code = 'i'
+    when LOG_WARN
+      code = 'w'
+    end
+
+    code
+  end
```
Comment (jmartin-r7) [inline lib/rex/logging/sinks/stream.rb:53]: Nice code consolidation. 👍

File: lib/rex/logging/sinks/flatfile.rb [hunk 2]
```diff
@@ -9,48 +9,16 @@ module Sinks
...
 
   #
   # Creates a flatfile log sink instance that will be configured to log to
   # the supplied file path.
   #
   def initialize(file)
-    self.fd = File.new(file, "a")
-  end
-
-  def cleanup # :nodoc:
```
Comment (jmartin-r7) [inline lib/rex/logging/sinks/flatfile.rb:24]: Should there still be a `cleanup` method to ensure the file is consistent if a file logger is replaced?  I do not see one in `Rex::Logging::Sinks::Stream`.

**Review by jmartin-tech — COMMENTED** [review]
Looks good, some minor thoughts noted.

**Comment by jmartin-r7** [comment]
I think some in progress PR work snuck into this, can you rebase on `master` and reduces just to your commits please?

File: lib/rex/logging/sinks/flatfile_without_colors.rb [hunk 3]
```diff
@@ -6,16 +6,17 @@ module Sinks
...
-class TimestampFlatfile < Flatfile
+class FlatfileWithoutColors < Flatfile
 
   def log(sev, src, level, msg) # :nodoc:
     return unless msg.present?
     msg = msg.gsub(/\x1b\[[0-9;]*[mG]/,'').gsub(/[\x01-\x02]/, ' ').gsub(/\s+$/,'')
-    fd.write("[#{get_current_timestamp}] #{msg}\n")
-    fd.flush
+    stream.write("[#{get_current_timestamp}] #{msg}\n")
+    stream.flush
```
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/rex/logging/sinks/flatfile_without_colors.rb:19]: Unfortunately we can't quite use `super` in this scenario. 

This logger was introduced within https://github.com/rapid7/metasploit-framework/pull/5573. Both the previous and current class names are misnomers. This logger is slightly different to the normal flatfile output. At a high level it:
- Adds timestamps
- Doesn't have additional severity/log level metadata such as `[d(0)] core` output that the `super` call would add
- Strips out colors
- Strips out trailing whitespace

This log sink is only used for session logs, and seems to roughly mimic the session I/O.

<details>

<summary>Example log file</summary>

```
[11/24/2020 17:34:27] 
[*] Logging started: 2020-11-24 17:34:27 +0000
[11/24/2020 17:34:52] load stdapi
[11/24/2020 17:34:52] load priv
[11/24/2020 17:42:38]   meterpreter   >
[11/24/2020 17:42:38] getsystem
[11/24/2020 17:42:39] ...got system via technique 1 (Named Pipe Impersonation (In Memory/Admin)).
[11/24/2020 17:42:43]   meterpreter   >
[11/24/2020 17:42:43] whoami
[11/24/2020 17:42:43] [-] Unknown command: whoami.
[11/24/2020 17:42:47]   meterpreter   >
[11/24/2020 17:42:47] getuid
[11/24/2020 17:42:48] Server username: NT AUTHORITY\SYSTEM
[11/24/2020 17:51:53]   meterpreter   >
[11/24/2020 17:51:53] dir
[11/24/2020 17:51:55] Listing: C:\Users\alan\Downloads
================================
Mode              Size     Type  Last modified              Name
----              ----     ----  -------------              ----
40777/rwxrwxrwx   0        dir   2020-09-21 18:08:17 +0100  DebugView
100666/rw-rw-rw-  1360262  fil   2020-09-21 18:08:03 +0100  DebugView.zip
40777/rwxrwxrwx   0        dir   2020-09-23 16:42:34 +0100  PipeList
100666/rw-rw-rw-  508191   fil   2020-09-23 16:42:24 +0100  PipeList.zip
100777/rwxrwxrwx  339096   fil   2020-06-19 16:26:08 +0100  PsExec.exe
100666/rw-rw-rw-  282      fil   2020-06-18 22:26:52 +0100  desktop.ini
100777/rwxrwxrwx  73802    fil   2020-12-02 16:01:46 +0000  reverse_windows (1).exe
100777/rwxrwxrwx  73802    fil   2020-12-02 16:01:47 +0000  reverse_windows (2).exe
100777/rwxrwxrwx  73802    fil   2020-12-02 16:01:30 +0000  reverse_windows.exe
[11/24/2020 17:51:57]   meterpreter   >
[11/24/2020 17:51:57] whoami
[11/24/2020 17:51:57] [-] Unknown command: whoami.
[11/24/2020 17:52:08]   meterpreter   >
[11/24/2020 17:52:08] whoami
[11/24/2020 17:52:08] [-] Unknown command: whoami.
[11/24/2020 17:52:11]   meterpreter   >
[11/24/2020 17:52:11] ls
[11/24/2020 17:52:12] Listing: C:\Users\alan\Downloads
================================
Mode              Size     Type  Last modified              Name
----              ----     ----  -------------              ----
40777/rwxrwxrwx   0        dir   2020-09-21 18:08:17 +0100  DebugView
100666/rw-rw-rw-  1360262  fil   2020-09-21 18:08:03 +0100  DebugView.zip
40777/rwxrwxrwx   0        dir   2020-09-23 16:42:34 +0100  PipeList
100666/rw-rw-rw-  508191   fil   2020-09-23 16:42:24 +0100  PipeList.zip
100777/rwxrwxrwx  339096   fil   2020-06-19 16:26:08 +0100  PsExec.exe
100666/rw-rw-rw-  282      fil   2020-06-18 22:26:52 +0100  desktop.ini
100777/rwxrwxrwx  73802    fil   2020-12-02 16:01:46 +0000  reverse_windows (1).exe
100777/rwxrwxrwx  73802    fil   2020-12-02 16:01:47 +0000  reverse_windows (2).exe
100777/rwxrwxrwx  73802    fil   2020-12-02 16:01:30 +0000  reverse_windows.exe
[11/24/2020 17:52:29]   meterpreter   >
[11/24/2020 17:52:29] 111111111
[11/24/2020 17:52:29] [-] Unknown command: 111111111.
[11/24/2020 17:52:43]   meterpreter   >
[11/24/2020 17:52:43] exit
[11/24/2020 17:52:43] [*] Shutting down Meterpreter...
[11/24/2020 17:52:45] 
[*] Logging stopped: 2020-11-24 17:52:45 +0000
```
</details>

If there's a better name to use for this log sink that is clearer to the next developer/user, let me know :+1:

[the remark this answers is not in the document]
Comment (jmartin-r7) [reply lib/rex/logging/sinks/flatfile_without_colors.rb:19]: Revised suggestion.
```
  def log(sev, src, level, msg) # :nodoc:
    return unless msg.present?
    if sev == LOG_RAW
      msg = msg.gsub(/\x1b\[[0-9;]*[mG]/,'').gsub(/[\x01-\x02]/, ' ').gsub(/\s+$/,'')
      msg = "[#{get_current_timestamp}] #{msg}"
    end
    super(sev, src, level, msg)
  end
```

This seems like an opportunity to become more consistent in logging. This method log takes `sev` and never uses it.  However looking closer this is called with `LOG_RAW` thru the following chain.
https://github.com/rapid7/metasploit-framework/blob/2cb2e38c49e8a896032811e5bcb71bbe2ba7297c/lib/rex/ui/text/shell.rb#L365-L377

https://github.com/rapid7/metasploit-framework/blob/2cb2e38c49e8a896032811e5bcb71bbe2ba7297c/lib/rex/logging/log_dispatcher.rb#L186-L192

Functionally the results should be the same, if we just inject the timestamp in the message at this level for `LOG_RAW`.

I agree another name would make sense though, the original intent was for all log lines even `LOG_RAW` to have timestamps.  This matched well with the original name except for the added stripping of color codes. `timestamp_colorless_flatfile.rb`?

[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/rex/logging/sinks/flatfile_without_colors.rb:19]: I'm not strongly opinionated here; I've changed to the suggested name and code - thanks! :+1:

**Comment by dwelch-r7** [comment]
# Release Notes

This PR adds a configuration option to the console and the rpc service for specifying the required metasploit logsink

**Pull request merged**